### ISSUE TITLE
Cannot reproduce "Bayesian Regression - Introduction (Part 1)"

### ISSUE THREAD

**Issue by leoguelman** [pyro-ppl/pyro#2047]
Hello, 

I'm getting non-sense predictions from the example in the Pyro site: <http://pyro.ai/examples/bayesian_regression.html?highlight=africa>

Here's what I'm running:

```python
import os
from functools import partial
import numpy as np
import pandas as pd
import seaborn as sns
import torch
import torch.nn as nn

import matplotlib.pyplot as plt

import pyro
from pyro.distributions import Normal, Uniform, Delta
from pyro.infer import SVI, Trace_ELBO
from pyro.optim import Adam
from pyro.distributions.util import logsumexp
from pyro.infer import EmpiricalMarginal, SVI, Trace_ELBO, TracePredictive
from pyro.infer.mcmc import MCMC, NUTS
import pyro.optim as optim
import pyro.poutine as poutine
from pyro.contrib.autoguide import AutoDiagonalNormal, AutoMultivariateNormal

# for CI testing
smoke_test = ('CI' in os.environ)
assert pyro.__version__.startswith('0.4.1')
pyro.enable_validation(True)
pyro.set_rng_seed(1)
pyro.enable_validation(True)


%matplotlib inline
#logging.basicConfig(format='%(message)s', level=logging.INFO)
# Enable validation checks
pyro.enable_validation(True)
smoke_test = ('CI' in os.environ)
pyro.set_rng_seed(1)
DATA_URL = "https://d2hg8soec8ck9v.cloudfront.net/datasets/rugged_data.csv"
data = pd.read_csv(DATA_URL, encoding="ISO-8859-1")
df = data[["cont_africa", "rugged", "rgdppc_2000"]]
#rugged_data = rugged_data[rugged_data["rgdppc_2000"].notnull()]
df = df[np.isfinite(df.rgdppc_2000)]
df["rgdppc_2000"] = np.log(df["rgdppc_2000"])

class RegressionModel(nn.Module):
    def __init__(self, p):
        # p = number of features
        super(RegressionModel, self).__init__()
        self.linear = nn.Linear(p, 1)
        self.factor = nn.Parameter(torch.tensor(1.))

    def forward(self, x):
        return self.linear(x) + (self.factor * x[:, 0] * x[:, 1]).unsqueeze(1)

p = 2  # number of features
regression_model = RegressionModel(p)

loc = torch.zeros(1, 1)
scale = torch.ones(1, 1)
# define a unit normal prior
prior = Normal(loc, scale)
# overload the parameters in the regression module with samples from the prior
lifted_module = pyro.random_module("regression_module", regression_model, prior)
# sample a nn from the prior
sampled_reg_model = lifted_module()

def model(x_data, y_data):
    # weight and bias priors
    w_prior = Normal(torch.zeros(1, 2), torch.ones(1, 2)).to_event(1)
    b_prior = Normal(torch.tensor([[8.]]), torch.tensor([[1000.]])).to_event(1)
    f_prior = Normal(0., 1.)
    priors = {'linear.weight': w_prior, 'linear.bias': b_prior, 'factor': f_prior}
    scale = pyro.sample("sigma", Uniform(0., 10.))
    # lift module parameters to random variables sampled from the priors
    lifted_module = pyro.random_module("module", regression_model, priors)
    # sample a nn (which also samples w and b)
    lifted_reg_model = lifted_module()
    with pyro.plate("map", len(x_data)):
        # run the nn forward on data
        prediction_mean = lifted_reg_model(x_data).squeeze(-1)
        # condition on the observed data
        pyro.sample("obs",
                    Normal(prediction_mean, scale),
                    obs=y_data)
        return prediction_mean

guide = AutoDiagonalNormal(model)

optim = Adam({"lr": 0.03})
svi = SVI(model, guide, optim, loss=Trace_ELBO(), num_samples=1000)

def train():
    pyro.clear_param_store()
    for j in range(num_iterations):
        # calculate the loss and take a gradient step
        loss = svi.step(x_data, y_data)
        if j % 100 == 0:
            print("[iteration %04d] loss: %.4f" % (j + 1, loss))

train()

get_marginal = lambda traces, sites:EmpiricalMarginal(traces, sites)._get_samples_and_weights()[0].detach().cpu().numpy()

def summary(traces, sites):
    marginal = get_marginal(traces, sites)
    site_stats = {}
    for i in range(marginal.shape[1]):
        site_name = sites[i]
        marginal_site = pd.DataFrame(marginal[:, i]).transpose()
        describe = partial(pd.Series.describe, percentiles=[.05, 0.25, 0.5, 0.75, 0.95])
        site_stats[site_name] = marginal_site.apply(describe, axis=1) \
            [["mean", "std", "5%", "25%", "50%", "75%", "95%"]]
    return site_stats

def wrapped_model(x_data, y_data):
    pyro.sample("prediction", Delta(model(x_data, y_data)))

posterior = svi.run(x_data, y_data)

trace_pred = TracePredictive(wrapped_model,
                             posterior,
                             num_samples=1000)

post_pred = trace_pred.run(x_data, None)

post_summary['prediction']['mean'].mean()
-362.3633981143727
```

I should be getting a positive number.

**Comment by fehiepsi** [comment]
I can confirm the issue. This is caused by recent changes in AutoGuide, which uses a new initialization `init_to_median` by default. There are a few solutions to this

#### solution 1: use `init_to_feasible` initialization
```
from pyro.infer.autoguide import init_to_feasible
guide = AutoDiagonalNormal(model, init_loc_fn=init_to_feasible)
```
#### solution 2: keep the default but change num_iterations from `1000` to `100,000`
#### solution 3: tuning other hyperparameters?

I would like to stick with solution 1 (because solution 2 will be very slow) but this needs us to explain what are those `init_loc_fn` functions. WDYT @fritzo @neerajprad ?

**Comment by leoguelman** [comment]
awesome! Solution 1 works.

**Comment by neerajprad** [comment]
Re-opening this issue since we should fix the initialization in the notebook. It seems odd to me that `init_to_median` is doing worse than `init_to_feasible` but I have seen this in some other cases. 

> I would like to stick with solution 1 (because solution 2 will be very slow) but this needs us to explain what are those init_loc_fn functions. WDYT @fritzo @neerajprad ?

Option 1 seems reasonable, but most likely we just got lucky with the initialization for that random seed, so I am not sure if we should rely on that. Another trick that @fritzo has mentioned is initializing the `auto_scale` parameter to a low value to aid convergence. I can take a look at this.

**Comment by leoguelman** [comment]
You are right. Looks like we were just lucky in the prior example. In this simple example, I've tried several initialization methods for the `guide` and none seem to work well in this simple problem. 

```python

import torch
import torch.nn as nn
import torch.nn.functional as F
import torch.optim as optim
import torchvision
from torchvision import datasets, transforms
from torch.utils.data import DataLoader

import pyro
from pyro.distributions import Normal, Uniform, Delta
from pyro.infer import SVI, Trace_ELBO
from pyro.optim import Adam
import pyro.poutine as poutine
from pyro.infer import EmpiricalMarginal, SVI, Trace_ELBO, TracePredictive
import pyro.optim as optim
from pyro.contrib.autoguide import AutoDiagonalNormal, AutoMultivariateNormal
from pyro.infer.autoguide import init_to_feasible, init_to_median

import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
from matplotlib import colors
from IPython import display
import os
from PIL import Image
from torch.utils.data.dataset import Dataset
from scipy.misc import imread

%matplotlib inline


### Simulate data

def f(x, sigma):
    epsilon = np.random.randn(*x.shape) * sigma
    return 10 * np.sin(2 * np.pi * (x)) + epsilon

train_size = 100
noise = 1.0

X = np.linspace(-0.5, 0.5, train_size).reshape(-1, 1)
y = f(X, sigma=noise)
y_true = f(X, sigma=0.0)

plt.scatter(X, y, marker='+', label='Training data')
plt.plot(X, y_true, label='Truth')
plt.title('Noisy training data and ground truth')
plt.legend();

# convert data to tensors
X = torch.tensor(X, dtype=torch.float)
y = torch.tensor(y, dtype=torch.float)


### Bayesian NN - Predicts just mean

# Build a custom nn module

class NN(nn.Module):
    
    def __init__(self, input_size, hidden_size1, hidden_size2, output_size):
        super(NN, self).__init__()
        self.linear_1 = nn.Linear(input_size, hidden_size1)
        self.linear_2 = nn.Linear(hidden_size1, hidden_size2)
        self.out = nn.Linear(hidden_size2, output_size)
        
    def forward(self, x):
        h1 = self.linear_1(x)
        h1_relu = F.relu(h1)
        h2 = self.linear_2(h1_relu)
        h2_relu = F.relu(h2)
        output = self.out(h2_relu)
        return output
    
# Set dimensions for units        
input_size, hidden_size1, hidden_size2, output_size = 1, 10, 10, 1   

# Construct our model by instantiating the class defined above
net = NN(input_size, hidden_size1, hidden_size2, output_size)

def model(x_data, y_data):
    
    w1_prior = Normal(loc=torch.zeros_like(net.linear_1.weight), scale=torch.ones_like(net.linear_1.weight))
    bias1_prior = Normal(loc=torch.zeros_like(net.linear_1.bias), scale=torch.ones_like(net.linear_1.bias))
    
    w2_prior = Normal(loc=torch.zeros_like(net.linear_2.weight), scale=torch.ones_like(net.linear_2.weight))
    bias2_prior = Normal(loc=torch.zeros_like(net.linear_2.bias), scale=torch.ones_like(net.linear_2.bias))
    
    wOut_prior = Normal(loc=torch.zeros_like(net.out.weight), scale=torch.ones_like(net.out.weight))
    biasOut_prior = Normal(loc=torch.zeros_like(net.out.bias), scale=torch.ones_like(net.out.bias))
    
    priors = {'linear_1.weight': w1_prior, 'linear_1.bias': bias1_prior, 
              'linear_2.weight': w2_prior, 'linear_2.bias': bias2_prior, 
              'out.weight': wOut_prior, 'out.bias': biasOut_prior}
    
    s_dev = pyro.sample("s_dev", Uniform(0., 5.))
    
    # lift module parameters to random variables sampled from the priors
    lifted_module = pyro.random_module("module", net, priors)
    
    # sample a nn from the prior
    lifted_reg_model = lifted_module()
    
    with pyro.plate("plate", len(x_data)):
        # run the nn forward on data
        prediction_mean = lifted_reg_model(x_data).squeeze(-1)
        # condition on the observed data
        pyro.sample("obs",
                    Normal(prediction_mean, s_dev),
                    obs=y_data)
        
        return prediction_mean

guide = AutoDiagonalNormal(model, init_loc_fn=init_to_median)


NUM_SAMPLES = 2000
NUM_ITERATIONS = 3000

optimizer = Adam({"lr": 0.03})

svi = SVI(model, 
          guide, 
          optimizer, 
          loss=Trace_ELBO(), 
          num_samples=NUM_SAMPLES)

pyro.clear_param_store()

losses = []
for i in range(NUM_ITERATIONS):
    loss = svi.step(X, y)
    losses.append(loss)
    if i % 100 == 0:
            print("[iteration %04d] loss: %.4f" % (i + 1, loss))


def get_predictions(X, n_samples=1000):
    
    preds = np.zeros(shape=(X.shape[0], n_samples))
    for i in range(n_samples):
        guide_trace = poutine.trace(guide).get_trace(X, None)
        reg = poutine.replay(model, guide_trace)
        preds[:, i] = reg(X, None).detach().cpu().numpy() 
    
    return preds
    
mean_pred = np.mean(preds, axis =1)  
quants = np.quantile(preds, q=[0.05,0.95], axis = 1)

plt.scatter(X, y, marker='+', label='Training data')
plt.plot(X, y_true, label='Truth')
plt.plot(X, mean_pred, marker='*', label='Predicted')
plt.fill_between((X.numpy()).reshape(X.shape[0],), 
                 quants[0,:], 
                 quants[1,:], 
                 alpha=0.5, label='90% prediction interval')
plt.title('Noisy training data and ground truth')
plt.legend();
```

Frequentist NN - Works ok

```


# Build a custom nn module

class NN(nn.Module):
    
    def __init__(self, input_size, hidden_size1, hidden_size2, output_size):
        super(NN, self).__init__()
        self.linear_1 = nn.Linear(input_size, hidden_size1)
        self.linear_2 = nn.Linear(hidden_size1, hidden_size2)
        self.out = nn.Linear(hidden_size2, output_size)
        
    def forward(self, x):
        h1 = self.linear_1(x)
        h1_relu = F.relu(h1)
        h2 = self.linear_2(h1_relu)
        h2_relu = F.relu(h2)
        output = self.out(h2_relu)
        return output
    
# Set dimensions for units        
input_size, hidden_size1, hidden_size2, output_size = 1, 10, 10, 1   

# Create random Tensors to hold inputs and outputs
#x = torch.randn(train_size, input_size)
#y = torch.randn(train_size, output_size)

# Construct our model by instantiating the class defined above
model = NN(input_size, hidden_size1, hidden_size2, output_size)

# Construct our loss function and an Optimizer. The call to model.parameters()
# in the SGD constructor will contain the learnable parameters of the two
# nn.Linear modules which are members of the model.
criterion = torch.nn.MSELoss(reduction='sum')
optimizer = torch.optim.SGD(model.parameters(), lr=1e-4)
for t in range(5000):
    # Forward pass: Compute predicted y by passing x to the model
    output = model(X)

    # Compute and print loss
    loss = criterion(output, y)
    if t % 100 == 99:
        print(t, loss.item())

    # Zero gradients, perform a backward pass, and update the weights.
    optimizer.zero_grad()
    loss.backward()
    optimizer.step()

plt.scatter(X, y, marker='+', label='Training data')
plt.plot(X, y_true, label='Truth')
plt.plot(X, output.detach().numpy(), marker='*', label='Predicted')
plt.title('Noisy training data and ground truth')
plt.legend();
```

**Comment by fritzo** [comment]
Note @wonyeol pointed out that guide's `sigma` distribution `Normal(...)` includes mass outside of the model's `sigma1` distribution (`Uniform(0,10)`), and this will occasionally error. I'm not sure if that's related to this issue, but we should probably switch to something consistent like `Exponential` and `LogNormal`.
